x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/utils.ts

```diff
@@ -193,6 +195,115 @@ export const getTupleDuplicateErrorsAndUniqueRules = (
   return [Array.from(errors.values()), Array.from(rulesAcc.values())];
 };
 
+const createQueryTerm = (input: string) => input.replace(/\\/g, '\\\\').replace(/\"/g, '\\"');
+const createQuery = (type: string, id: string, rawIdPrefix: string) =>
+  `"${createQueryTerm(`${rawIdPrefix}${type}:${id}`)}" | "${createQueryTerm(id)}"`;
+
+/**
+ * Query for a saved object with a given origin id and replace the
+ * id in the provided action with the _id from the query result
+ * @param action
+ * @param esClient
+ * @returns
+ */
+export const swapActionIds = async (
+  action: Action,
+  savedObjectsClient: SavedObjectsClientContract
+): Promise<Action | Error> => {
+  try {
+    // TODO: might need to pass down namespace
+    // but for now it's blank to do local testing.
+    const search = createQuery('action', action.id, '');
+    const foundAction = await savedObjectsClient.find<Action>({
+      type: 'action',
+      search,
+      rootSearchFields: ['_id', 'originId'],
+    });
+
+    if (foundAction.saved_objects.length === 1) {
+      // id is of the form 'action:1234-5678...' so I need to split off the 'action' prefix
+      return { ...action, id: foundAction.saved_objects[0].id };
```

Comment:
References should always use the exact IDs; we want to reduce reliance on `resolve` as much as possible because legacy URL aliases !== shared object origin.
(you can have a shared origin without a legacy URL alias)